Add release helper script

Rumperuu opened the pull request:
This PR adds a release helper script, amongst other workflow improvements.

See `README.md` for instructions, and `_tools/release.sh` for the script itself.

This script:

1. sanity-checks the various version tags;
2. triggers a Plugin build;
3. flags the new version as pre-release;
4. tags the version in Git;
5. creates a local working copy of the SVN repo;
6. copies the new release to the local `trunk/` (whilst keeping the ‘Stable Tag’ field in `readme.txt` pointing to the previous stable version);
7. copies the commit message from the changelog in `readme.txt`; and
8. (if a flag is set) commits the changes to the remote `trunk/` and copies them to `tags/<new pre-release version>`.

At various stages user input is required to validate information. This is not ready for automation with GitHub Actions, but is a useful step on the way — see [this piece](https://blog.danslimmon.com/2019/07/15/do-nothing-scripting-the-key-to-gradual-automation/) for more info.

Unless a `-c` flag is passed (e.g., by running `composer run release:commit`) no changes will take place on the remote SVN repo. If you want to test this out on a branch other than `main`, see lines 31 & 52 of the script.

Version checking enforces the versioning rules available [here](https://github.com/markcheret/footnotes/wiki/Versioning).

**NB: I have not tested the `-c` mode yet, as I wanted people will more familiarity with the SVN to have a look at it before I risked making any changes and blowing everything up.**

Closes #5, #30 and #40

pewgeuges reviewed (approved):
Everything fixed and reviewed at this point, but need to test how everything works out.

Testing on SVN is IMO not something anybody can do while we didn’t get this sorted out.

Rumperuu commented on readme.txt:
```diff
@@ -4,10 +4,7 @@ Tags: footnote, footnotes, bibliography, formatting, notes, Post, posts, referen
 Requires at least: 3.9
 Tested up to: 5.7
 Requires PHP: 5.6
-Package Version: 2.5.11d2
-Version: 2.5.10
 Stable Tag: 2.5.10
-CAUTION: THE S. T. FIELD IS PARSED FOR RELEASE CONFIGURATION.
```
They were useful before, but they are no longer needed

pewgeuges replied:
The users may be reassured at the sight of these version numbers.

Moreover we need them as designed.

Rumperuu commented on _tools/build.sh:
```diff
@@ -1,9 +1,24 @@
+
+# Moves everything else over to `dist/`
+echo "Copying directories..."
+cp -r -t dist class/ js/ img/ languages/ templates/
```
Minifying files is used to reduce the file sizes, so distributing both minified and un-minified versions of all the files doesn't make a lot of sense.

Rumperuu replied:
Also, on the opposite note, I've re-added the minified CSS files to `css/` for now so that the cache-busting with `filemtime()` won't fail when testing the dev codebase (i.e., when the minified files are in `dist/css/` rather than root `css/`. I need to add a prod/dev flag that'll set the correct path for the stylesheet enqueues, but this bodge job will work for now.

pewgeuges replied:
> Minifying files is used to reduce the file sizes, so distributing both minified and un-minified versions of all the CSS files would be pretty self-defeating.

Open source projects are to distribute the unminified stylesheets and scripts for referernce. But not the concatenated sources of the 12 minified style sheets used (plus settings.css).

pewgeuges replied:
> Also, on the opposite note, I've re-added the minified CSS files to `css/` for now so that the cache-busting with `filemtime()` won't fail when testing the dev codebase (i.e., when the minified files are in `dist/css/` rather than root `css/`. I need to add a prod/dev flag that'll set the correct path for the stylesheet enqueues, but this bodge job will work for now.

Developers can use a plugin for cache busting, don’t need to implement this, anyway useless when making changes to test (unminified style sheets when the production mode Boolean is set to false).

How are you going to use the WordPress URL parameter scheme?

That’s why this constant was added in the first place.

Rumperuu replied:
> anyway useless when making changes to test (unminified style sheets when the production mode Boolean is set to false).

Good shout, I've added a toggle for that flag to the stylesheet build command in 7d21dd9 

> How are you going to use the WordPress URL parameter scheme?

Can you clarify what you're talking about? AFAICT the constant was only used to append a version number to the end of the stylesheet imports, so that a new dev version == a new tag at the end of the stylesheet enqueue == bypassing the cache.

However, having this be based on a manually-updated constant is not a great solution. Instead, we can use `filemtime()` to append a URL query parameter based on when the stylesheet was last modified. This means we any updates to stylesheets will be reflected without anyone having to increment a dev version.

> > Minifying files is used to reduce the file sizes, so distributing both minified and un-minified versions of all the CSS files would be pretty self-defeating.
> 
> Open source projects are to distribute the unminified stylesheets and scripts for reference. But not the concatenated sources of the 12 minified style sheets used (plus settings.css).

We fulfil the requirements of an open source project by providing the unminified stylesheets here. Distributing both minified and unminified files is unnecessary, but at the end of the entire `css/` directory (including base, concatenated and minified stylesheets) is only 340 kB so I'm not overly bothered either way. To be fair, if you wanted to distribute the unminified files so that users can explore them themselves, our footprint is small enough that I would say we don't really need to bother minifying them in the first place. Whether the stylesheets are 240.3 kB (unminified concatenated) or 67 kB (minified concatenated) is barely going to make a difference to anyone.

pewgeuges replied:
> > anyway useless when making changes to test (unminified style sheets when the production mode Boolean is set to false).
> 
> Good shout, I've added a toggle for that flag to the stylesheet build command in [7d21dd9](https://github.com/markcheret/footnotes/commit/7d21dd9d1eec0929429ed22090d3632aa03ddd87)

As noted there, it’s set the wrong way.

> 
> > How are you going to use the WordPress URL parameter scheme?
> 
> Can you clarify what you're talking about?

The 4th parameter of wp_enqueue_style().

See https://developer.wordpress.org/reference/functions/wp_enqueue_style/

> AFAICT the constant was only used to append a version number to the end of the stylesheet imports, so that a new dev version == a new tag at the end of the stylesheet enqueue == bypassing the cache.

No.

> 
> However, having this be based on a manually-updated constant is not a great solution. Instead, we can use `filemtime()` to append a URL query parameter based on when the stylesheet was last modified. This means we any updates to stylesheets will be reflected without anyone having to increment a dev version.

The constant is needed  anyway for the visitors of pages using Footnotes.

> 
> > > Minifying files is used to reduce the file sizes, so distributing both minified and un-minified versions of all the CSS files would be pretty self-defeating.
> > 
> > 
> > Open source projects are to distribute the unminified stylesheets and scripts for reference. But not the concatenated sources of the 12 minified style sheets used (plus settings.css).
> 
> We fulfil the requirements of an open source project by providing the unminified stylesheets here.

The unminified files are distributed alongside the minified files. We can skip the concatenated ones, though.

> 
> Distributing both minified and unminified files is unnecessary, but at the end of the day the entire `css/` directory (including base, concatenated and minified stylesheets) is only 340 kB so I'm not overly bothered either way.
> 
> That said, if you wanted to distribute the unminified files so that users can explore them themselves, our footprint is small enough that I would say we don't _really_ need to bother minifying them in the first place. Whether the stylesheets are 240.3 kB (unminified concatenated) or 67 kB (minified concatenated) is barely going to make a difference to anyone.

See https://wordpress.org/support/topic/simply-speed-optimisation/

Rumperuu replied:
>>> How are you going to use the WordPress URL parameter scheme?
>>Can you clarify what you're talking about?
>The 4th parameter of wp_enqueue_style().
>
>See https://developer.wordpress.org/reference/functions/wp_enqueue_style/

Thought so. See a66faae — former arguments of `C_STR_PACKAGE_VERSION` have been replaced with calls to `filemtime( plugin_dir_path( dirname( __FILE__ ) ) . 'css/<filename>.css' )`

> See https://wordpress.org/support/topic/simply-speed-optimisation/

Fair enough

pewgeuges commented:
> Version checking enforces the versioning rules available [here](https://github.com/markcheret/footnotes/wiki/Versioning).

I don’t know where this versioning scheme comes from, but I do know that the d series are delta versions (or development versions) and they’re numbered zero-based.

Pre-releases are not tagged on SVN, nor are tags renamed. WordPress directs us to host the latest version in trunk, even if a beta, see [here](https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#use-the-trunk-folder-for-code), and to make tags when releasing.

Moreover, we need to be careful to maintain traceability in SVN, by committing as many d versions as needed for that purpose.

Rumperuu commented on footnotes.php:
```diff
@@ -26,7 +24,7 @@
  * @var str
  * @lastmodified 2021-03-11T2202+0100
  */
-define( 'C_STR_PACKAGE_VERSION', '2.5.11d2' );
```
We can do one better and use proper file modification-based cache-busting, letting us get rid of this constant altogether! Done in a66faae

Rumperuu commented on readme.txt:
```diff
@@ -81,9 +78,6 @@ Visit this swift write-up from a **footnotes** user by the name of **Southwest**
 == Changelog ==
 
 = 2.5.11 =
-- Adding: Documentation: Readme.txt: comment line below the 'Stable Tag' field to warn that this is (unexpectedly) parsed for release configuration.
-- Adding: Documentation: Readme.txt: informative 'Version' field pointing a tag folder in SVN for bugfix versions available ahead of the stable tag.
-- Adding: Documentation: Readme.txt: informative 'Package Version' field in sync with the package version currently hidden in the main PHP script header.
```
State-of-the-art doesn't mean filling headers with custom tags. Once this PR is merged, there will be script-enforced version number constraints and so we won't have any more dramas like the other week.

pewgeuges replied:
> State-of-the-art doesn't mean filling headers with custom tags. Once this PR is merged, there will be script-enforced version number constraints and so we won't have any more dramas like the other week and won't need warnings in the header.

When I wrote “state-of-the-art” I neither referred to custom tags, nor to filling headers with them.

State-of-the-art means that every readme has a version number for the actual version. When WordPress has two configuration version fields (one in readme, another one in main.php), these add to that.

As of the warning, at this point it is designed to reassure Footnotes users.

Rumperuu commented:
> > Version checking enforces the versioning rules available [here](https://github.com/markcheret/footnotes/wiki/Versioning).
> 
> I don’t know where this versioning scheme comes from, but I do know that the d series are delta versions (or development versions) and they’re numbered zero-based.

They can be, but they don't need to be. As they are no longer needed for stylesheet cache-busting (see a66faae), maintaining multiple development releases adds developer overhead and the likelihood for mistakes for no benefit.

> Pre-releases are not tagged on SVN, nor are tags renamed. WordPress directs us to host the latest version in trunk, even if a beta, see [here](https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#use-the-trunk-folder-for-code), and to make tags when releasing.

Not quite. That link says that `trunk/` can include a beta codebase, but it doesn't say anything about `tags/` being limited to stable codebases one way or another. In that case, the use of `p` versions is up to personal preference; functionally, there's no difference between having `trunk/` contain the `p` codebase and having `tags/<version>p/` contain it, provided `trunk/readme.txt`'s ‘Stable Tag’ points to the latest non-`p` version.

Although, in retrospect, I think the former is preferable. On the ‘Previous versions’ selector `trunk/` already appears as ‘Development Version’, plus there's no real point in having `trunk/` and a `tags/<version>p/` with identical code. I've removed the step in 74b9ea3 and updated [the Wiki page](https://github.com/markcheret/footnotes/wiki/Versioning) to reflect this.

> Moreover, we need to be careful to maintain traceability in SVN, by committing as many d versions as needed for that purpose.

The SVN tracks the project history at a lower level of granularity to the Git repo, as it's a release system. Git branch commits are the highest level of detail (e.g., ‘Replace variable string with proper stylesheet cache-busting’ from a66faae), which are then squashed into a higher-level summary when merging into `main`. On release, all of these squash commits are then summarised again and entered into the changelog in `readme.txt` and the SVN commit.

In an ideal world, the SVN repo. will have one commit per version. In reality, this won't be the case if we ever have to push a pre-release bugfix, but it's a good rule of thumb.

pewgeuges reviewed (dismissed):
I don’t agree with the latest changes, see individual comments.

Rumperuu commented:
Reverted to draft because I briefly thought this would break on one of prod or dev due to incorrect stylesheet filepaths, then remembered I'd temporarily fixed it with 4109bbc :relieved:

pewgeuges commented:
> > > Version checking enforces the versioning rules available [here](https://github.com/markcheret/footnotes/wiki/Versioning).
> > 
> > 
> > I don’t know where this versioning scheme comes from, but I do know that the d series are delta versions (or development versions) and they’re numbered zero-based.
> 
> They can be, but they don't need to be. As they are no longer needed for stylesheet cache-busting (see [a66faae](https://github.com/markcheret/footnotes/commit/a66faae03d720ea4dadf84b4f1bcf548e9c92502)), maintaining multiple development releases adds developer overhead and the likelihood of errors for no benefit. In the bin they go.

Users need up-to-date version URL parameters for cache busting.

> 
> > Pre-releases are not tagged on SVN, nor are tags renamed. WordPress directs us to host the latest version in trunk, even if a beta, see [here](https://developer.wordpress.org/plugins/wordpress-org/how-to-use-subversion/#use-the-trunk-folder-for-code), and to make tags when releasing.
> 
> Not quite. That link says that `trunk/` can include a beta codebase, but it doesn't say anything about `tags/` being limited to stable codebases one way or another. In that case, the use of `p` versions is up to personal preference; functionally, there's no difference between having `trunk/` contain the `p` codebase and having `tags/<version>p/` contain it (providing that in both cases, `trunk/readme.txt`'s ‘Stable Tag’ points to the latest non-`p` version).

We only need tags for actual releases.

> 
> Although, in retrospect, I think the former is preferable. On the [‘Previous versions’ selector](https://wordpress.org/plugins/footnotes/advanced/) (which is how users would access the pre-release versions anyway) `trunk/` already appears as ‘Development Version’, plus there's no real point in having `trunk/` and a `tags/<version>p/` with identical code. I've removed the step in [74b9ea3](https://github.com/markcheret/footnotes/commit/74b9ea386f7d0b3a96f5ffc76ad8a01d8c16d5af) and updated [the Wiki page](https://github.com/markcheret/footnotes/wiki/Versioning) to reflect this.

Thanks.

> 
> > Moreover, we need to be careful to maintain traceability in SVN, by committing as many d versions as needed for that purpose.
> 
> The SVN tracks the project history at a lower level of granularity to the Git repo, as it's a release system. Git branch commits are the highest level of detail (e.g., ‘Replace variable string with proper stylesheet cache-busting’ from [a66faae](https://github.com/markcheret/footnotes/commit/a66faae03d720ea4dadf84b4f1bcf548e9c92502)), which are then squashed into a higher-level summary of changes when merging into `main`. On release, all of these squash commits are then summarised again and entered into the changelog in `readme.txt` and the SVN commit.

Traceability commits to SVN are useful also before and after moving blocks of code.

> 
> In an ideal world, the SVN repo. will have one commit per version. In reality, this won't be the case if we ever have to push a pre-release bugfix, but it's a good rule of thumb.

Footnotes did have such one-commit-per-version releases lastly.

lolzim reviewed (approved):
Nice pack!

markcheret reviewed (approved):
Sweet deal. Thanks for your work. Can't wait for auto-releases. Wouldn't wanna do them manually :)

Rumperuu commented on _tools/build-stylesheets.sh:
```diff
@@ -60,6 +60,14 @@ elif [[ $1 == "-d" ]]; then
+		cp $f ./dist/css
+		#mv $f ./dist/css
+		echo -e '\t' $filename".css moved."
+	done
```
@pewgeuges Re-added the unminified stylesheets to `dist/css/`. If you wanted to include the development stylesheets instead (e.g., `dev-common.css`) rather than the concatenated ones, just tweak line 65 to say `for f in ./css/*.css; do`

Rumperuu replied:
Changed in bc65b19 and f693684 to deploy `dev-` stylesheets alongside `.min.css` ones.

Rumperuu commented:
Will merge with `main` at 16:00 today (UK time) if there are no additional objections by then.

markcheret reviewed (commented):
Looking sleek
